File: backend/app/models/attendance.py
"""Attendance model for tutoring sessions."""

from __future__ import annotations

import enum
import uuid
from datetime import datetime

from sqlalchemy import DateTime, Enum, ForeignKey, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class AttendanceStatus(str, enum.Enum):
    PRESENT = "PRESENT"
    ABSENT = "ABSENT"
    JUSTIFIED = "JUSTIFIED"


class SessionAttendance(Base):
    """Attendance per tutorando in a session."""

    __tablename__ = "session_attendance"

    session_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("sessions.id", ondelete="CASCADE"),
        primary_key=True,
    )
    tutorando_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("users.id", ondelete="CASCADE"),
        primary_key=True,
    )
    status: Mapped[AttendanceStatus] = mapped_column(Enum(AttendanceStatus, name="attendancestatus"), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        onupdate=func.now(),
    )

    session = relationship("Session")
    tutorando = relationship("User")

File: frontend/src/types/attendance.ts
import type { AdminUser } from "./users";
import type { SessionItem } from "./sessions";

export type AttendanceStatus = "PRESENT" | "ABSENT" | "JUSTIFIED";

export interface AttendanceEntry {
  tutorando: AdminUser;
  status?: AttendanceStatus | null;
  updated_at?: string | null;
}

export interface AttendanceUpdateEntry {
  tutorando_id: string;
  status: AttendanceStatus;
}

export interface TutorandoAttendanceItem {
  session: SessionItem;
  status?: AttendanceStatus | null;
  updated_at?: string | null;
}

export interface AttendanceUpdatePayload {
  entries: AttendanceUpdateEntry[];
}
